### PR TITLE
App default creds

### PR DESCRIPTION
**Pull request by mgilson**
Only authenticate individual requests in a batch if the credentials are non-`None`.  Otherwise, assume that an appropriate authentication is being provided at a higher level (e.g. for the entire batch)

This should address #350

### CODE REVIEW

File: googleapiclient/http.py [hunk 1]
```diff
@@ -1202,7 +1202,17 @@ def _serialize_request(self, request):
 
     if request.http is not None and hasattr(request.http.request,
         'credentials'):
-      request.http.request.credentials.apply(headers)
+      credentials = request.http.request.credentials
```
Comment (mgilson) [inline googleapiclient/http.py]: I've played around with a couple different implementations here and the tests have led me to recognize a few other corner cases and/or places where this isn't ideal.  Specifically, 

* `Credentials` doesn't have an `access_token` property.
* `OAuth2Credentials` does have an `access_token` property.
* `AssertionCredentials` (which eventually subclass `OAuth2Credentials` _do_ have an `access_token` property, but it gets set to `None` in `__init__` which leaves `AssertionCredentials` with a non-working `apply` method
  - `apply` looks to be inherited from `OAuth2Credentials` which assumes that `access_token` is a string

I'm not super familiar with the architecture of this library and the architecture of the `oauth2client` library and how much their requirements are supposed to interoperate.  Having looked at this more, it feels like the bug might actually be in `oauth2client.client.AssertionCredentials.apply`.  If that method were to become a no-op (at least if `self.access_token is None`), then the problems _here_ would go away.  I'm not sure if that's the right thing to do _there_ though.

Maybe this isn't the right place to raise that issue (but it looks like a lot of the committers to both libraries are the same) -- and I suppose the fix here would still need to be in place to handle interoperability with past versions of that library.  Anyway, any advice you can give on this front is welcome :-)

**Comment by jonparrott** [comment]
@mgilson thanks for doing this- heads up I'm pretty busy the next couple of days but I will try to get to this by friday.

**Comment by mgilson** [comment]
Thanks for letting me know when to be looking into this again.  I'm in no hurry -- We've already coded the workaround that I mentioned in the issue on our end so there's nothing mission critical for us.  In other words, take your time.  :-)

**Comment by jonparrott** [comment]
@mgilson sorry I got super backlogged on this. I think it might not be as easy as we think - this library has to support both oauth2client and google-auth. We might have to add some special logic to handle those. Let me find some time to look into this in more depth to see what's up.

**Comment by jonparrott** [comment]
Closing in favor of #376.
